Sanitize the filesystem data model and query.

johnworth opened the pull request:
## Scope
Move the folder listing return value into the Folder type and rename the top-level query.

## Changes

- Moves listing into the Folder type and renames it 'contents'.
- Removes the top-level listFolder query.
- Modifies the resolvers to handle the new layout.
- Modifies TerrainDataSource to remove stat information from the listFolder return value, which is now an iterable.

## Problems
Nesting the folder listings gets expensive quickly. 

## Sample query containing nested listings
The return value is too large to reproduce here.

```
{
  filesystem(paths: [
    "/iplant/home/ipcdev/"
  ]) {
    id
    name: label
    path
    type
    
    ... on File {
      fileSize
    }
    
    ... on Folder {
      contents {
        id
        name: label
        path
        type
        
        ... on File {
          fileSize
        }
        
        ... on Folder {
          contents {
            id
            name: label
            path
            type
          }
        }
      }
    }
  }
}
```

slr71 commented on src/server/graphql/resolvers/Filesystem.js:
```diff
@@ -44,17 +44,30 @@ export default {
+        },
+        shareCount: async (folder, _args, { dataSources }) => {
+            const stat = await dataSources.terrain.filesystemStat(folder.path);
+            return stat["share-count"];
```
Does GraphQL manage caching the result of the call to the `stat` endpoint for us?

johnworth replied:
The caching provided by Apollo doesn't provide batching as well, AFAIK. We're both batching requests and caching it with dataloader. Not sure if the in-process LRU cache would be okay in our environment because of the multiple replicas and I'd really rather not add Redis or memcached to the stack until we have multiple places that absolutely require one of them.

ianmcorvidae replied:
Doing my thing and reading about it, I guess Dataloader caches within the context of a single request, but not cross-request. Still a nice improvement, probably, and it's really hard for us to provide nice cache semantics outside of a single request. The caching provided by Apollo might be more useful in non-filesystem contexts where we can provide more useful cache headers in Terrain if we want (of course, it'll also be less useful there, but anyway).

psarando commented on src/server/graphql/typesDefs/Filesystem.js:
```diff
@@ -113,6 +118,11 @@ export default gql`
         fileCount: Int
+
+        """
+        This an expensive field to include inside content listings.
```
👍 good comments

psarando commented on src/server/graphql/typesDefs/Filesystem.js:
```diff
@@ -98,16 +105,25 @@ export default gql`
+        """
+        This an expensive field to include inside content listings.
+        """
+        folderCount: Int
```
I almost forgot: the file and folder counts are slow because they come from the stat endpoint, but the listing endpoint does return a `total` field (the combined count for the listing), and that will be required for paging listings in the UI.

johnworth commented:
folderCount and contentType are now available. 

I'll go ahead and convert over to ES6 syntax now, I'd rather standardize on that now and I don't want to forget about it later.

johnworth commented on src/server/graphql/dataSources/TerrainDataSource.js:
```diff
@@ -41,12 +42,16 @@ export default class TerrainDataSource extends RESTDataSource {
-    }
+
+        // Guarantees ordering, hopefully
+        return _.map(fullPaths, (path) => resp.paths[path]);
```
Should be fixed now, thanks.

johnworth commented:
I added the `FileContents` type and changed `File.contents` to it. It contains `total`, `totalBad`, and `listing` fields. A client query to grab that information could look like this:

```
{
  filesystem(path: "/iplant/home/ipcdev/testing-vice") {
    ... on Folder {      
      contents {
        total
        totalBad
        
        listing {
          path
          type
        }
      }
    }
  }
}
```

psarando commented on src/server/graphql/dataSources/TerrainDataSource.js:
```diff
@@ -77,34 +81,24 @@ export default class TerrainDataSource extends RESTDataSource {
+        });
+
+        return {
+            total,
```
Don't forget to change the `@returns` JSDoc for this function, too.

sri2k1us commented:
Yes! this `_` syntax is driving me nuts 😩.
> I'd also still like to explore getting rid of `lodash` in favor of ES6 syntax, but that can be saved for the next PR.

sri2k1us commented on src/server/graphql/dataSources/TerrainDataSource.js:
```diff
@@ -82,24 +81,20 @@ export default class TerrainDataSource extends RESTDataSource {
-        return listing;
+        return [...listingFolders, ...listingFiles];
     }
 }
```
👍
